### PR TITLE
Strengthen type capture for specialization

### PR DESCRIPTION
**Pull request by vmarcos**
This PR strengthens the type capture step for arrangement specialization to ensure that input `types` in `ArrangeBy` are only set to `Some` if the `enable_specialized_arrangements` flag is turned on. Additionally, the PR strengthens the checks when using type information to only proceed with specialization when the type capture is not out-of-sync with the flag.

This PR builds on the design introduced by #21968.

### Motivation

  * This PR fixes a recognized bug. Fixes https://github.com/MaterializeInc/materialize/issues/22778

### Tips for reviewer

In addition to strengthening the checks, a regression test is provided as part of the `cluster` tests. Moreover, the strengthening of type capture was carried through to planning of `ArrangeBy` in `Threshold` lowering, as a special case is handled there to plan a required arrangement when necessary.

The diff is easier to read if you hide whitespace.

[Slack ref](https://materializeinc.slack.com/archives/C02PPB50ZHS/p1698666467029299)

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Review by antiguru — APPROVED** [review]
I think this looks fine, but I left some comments around documentation and defensive asserts, and some future work we should do. Thank you!

File: src/compute-types/src/plan/threshold.rs [hunk 1]
```diff
@@ -108,10 +109,10 @@ impl ThresholdPlan {
     /// This is likely either an empty vector, for no arrangement,
     /// or a singleton vector containing the list of expressions
     /// that key a single arrangement.
-    pub fn keys(&self) -> AvailableCollections {
+    pub fn keys(&self, types: Option<Vec<ColumnType>>) -> AvailableCollections {
```
[the remark this answers is not in the document]
Comment (vmarcos) [reply src/compute-types/src/plan/threshold.rs:114]: Added!

File: src/compute-types/src/plan/mod.rs [hunk 2]
```diff
@@ -159,6 +159,7 @@ impl AvailableCollections {
     /// specified ways.
     pub fn new_arranged(
         arranged: Vec<(Vec<MirScalarExpr>, BTreeMap<usize, usize>, Vec<usize>)>,
+        types: Option<Vec<ColumnType>>,
```
[the remark this answers is not in the document]
Comment (vmarcos) [reply src/compute-types/src/plan/mod.rs:163]: Done!

File: src/compute/src/render/context.rs [hunk 3]
```diff
@@ -1054,28 +1054,40 @@ where
...
-                "FormArrangementKey",
-                specialized_arrangement_key(key.clone(), thinning.clone(), None, None),
-            );
-            let oks = oks.mz_arrange::<RowSpine<Row, Row, _, _>>(name);
-            (SpecializedArrangement::RowRow(oks), errs)
+        if enable_specialized_arrangements {
+            // We must have captured types if the flag is on. However, if the flag value
+            // and the available types fell out-of-sync, we will just skip specialization
+            // and use the RowRow catch-all below.
+            if let (Some(_key_types), Some(val_types)) = (key_types, val_types) {
```
[the remark this answers is not in the document]
Comment (vmarcos) [reply src/compute/src/render/context.rs]: Oh, interesting question, which reveals that the types here are not capturing the promised semantics here in its entirety. I changed the return type of `derive_key_val_types` and the input types here to `Option<(Vec<ColumnType>, Vec<ColumnType>)>`.

In general, the logic is that: (a) We capture only input types to an `ArrangeBy` operator, when necessary; and (b) From the input types, desired key, and thinning, we derive both key and value types at once.

File: src/compute-types/src/plan/mod.rs [hunk 4]
```diff
@@ -1012,6 +1013,9 @@ impl<T: timely::progress::Timestamp> Plan<T> {
             forms.arranged.extend(collections.arranged);
             forms.arranged.sort_by(|k1, k2| k1.0.cmp(&k2.0));
             forms.arranged.dedup_by(|k1, k2| k1.0 == k2.0);
+            if forms.types.is_none() {
+                forms.types = collections.types;
+            }
```
[the remark this answers is not in the document]
Comment (vmarcos) [reply src/compute-types/src/plan/mod.rs]: The function is adopting a merge semantics, so it's OK also if `collections.types.is_none()` even if `forms.types.is_some()`. But I agree that if we performed type capture for the input for both `forms` and `collections`, then they should agree. So I added an `assert!` now capturing the two possibilities.

**Comment by vmarcos** [comment]
TFTR!

**Pull request merged**